src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -501,9 +515,39 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
             return this.GetTriviaData(tokenData1, tokenData2).SecondTokenIsFirstTokenOnLine;
         }
 
+        /// <summary>
+        /// Gets approximated index bounds for a token with specified position, which will definitely contain
+        /// the token, if it is present in the stream.
+        /// </summary>
+        private void GetTokenIndexBounds(int position, out int lowerBound, out int upperBound)
+        {
+            float avgLength = this.AverageTokenLength;
+            Debug.Assert(avgLength > 0);
+            // make an initial guess about the token position
+            lowerBound = upperBound = Bounded((int)((position - _tokens[0].FullSpan.Start) / avgLength), min: 0, max: _tokens.Length - 1);
+
+            // decrease lower bound until the desired position is strictly after it
+            while (_tokens[lowerBound].FullSpan.Start >= position && lowerBound > 0)
+            {
+                int estimatedError = (int)((_tokens[lowerBound].FullSpan.Start - position) / avgLength);
+                lowerBound = Math.Max(0, lowerBound - Math.Max(1, estimatedError));
+            }
+
+            // increase upper bound until the desired position is strictly before it
+            while (_tokens[upperBound].FullSpan.Start <= position && upperBound < _tokens.Length - 1)
```

Comment:
shouldn't this be the *end* of the upper bound, not the *start*?

Reply:
No, it should not, since we are looking for start of the token

Reply:
Maybe renaming `position` to `start` or `tokenStart` would clarify that

Reply:
this also needs to be documented in the code.  right now it looks like a bug (and i'm still not actually understandin the explanation).

Reply:
ok.... i *think* i'm getting it.  but yes, 'position' is vague.  naming to tokenFullStart might be better.  'start' has a different meaning, so i wouldn't use that.

Reply:
Renamed to `tokenStart`